### PR TITLE
Audio events

### PR DESCRIPTION
**Pull request by NickM-27**
The initial implementation of the audio events feature is going to be minimal. The audio model runs well on CPU so that will be the only option.

To-Do:

- [x] Cleanup labelmap
- [x] Add ability to enable / disable audio events via MQTT
- [x] Optimize stop event logic

### CODE REVIEW

File: frigate/config.py [hunk 1]
```diff
@@ -795,6 +814,11 @@ def verify_config_roles(camera_config: CameraConfig) -> None:
             f"Camera {camera_config.name} has rtmp enabled, but rtmp is not assigned to an input."
         )
 
+    if camera_config.audio.enabled and "audio" not in assigned_roles:
+        raise ValueError(
+            f"Camera {camera_config.name} has audio events enabled, but audio is not assigned to an input."
+        )
```
Comment (skrashevich) [inline frigate/config.py:820]: It feels like one of this settings is superfluous and unnecessary

File: frigate/const.py [hunk 2]
```diff
@@ -12,6 +12,13 @@
...
+# Audio Consts
+
+AUDIO_DETECTOR = "audio-detector"
+AUDIO_DURATION = 0.975
+AUDIO_FORMAT = "s16le"
+AUDIO_SAMPLE_RATE = 16000
+
 # Regex Consts
 
 REGEX_CAMERA_NAME = r"^[a-zA-Z0-9_-]+$"
```
Comment (skrashevich) [inline frigate/const.py:24]: `REGEX_CAMERA_NAME = r"^[\w-]+$"`

File: frigate/events/audio.py [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+                [i.path for i in self.config.ffmpeg.inputs if "audio" in i.roles][0],
+                self.pipe,
+            )
+        )
+        self.pipe_file = None
+        self.logpipe = LogPipe(f"ffmpeg.{self.config.name}.audio")
+        self.audio_listener = None
+
+    def detect_audio(self, audio) -> None:
+        waveform = (audio / 32768.0).astype(np.float32)
```
Comment (skrashevich) [inline frigate/events/audio.py:149]: > 32768.0

magic number?

File: frigate/config.py (hunk 1 above)
Comment (NickM-27) [reply frigate/config.py:820]: `audio` must be assigned to a specific role, we can not assume which role it is assigned to. This is in line with how the other roles in frigate works.

File: frigate/const.py (hunk 2 above)
Comment (NickM-27) [reply frigate/const.py:24]: That has nothing to do with this PR so should be handled outside of it

File: frigate/config.py (hunk 1 above)
Comment (skrashevich) [reply frigate/config.py:820]: other roles doesn't requires second setting at camera level. for example, if you add "record" role to input, you doesn't need to additionaly enable recording in camera level

File: frigate/events/audio.py (hunk 3 above)
Comment (NickM-27) [reply frigate/events/audio.py:149]: https://stackoverflow.com/questions/59658218/please-explain-why-we-use-32768-in-audio-signals

I will update to use a const

File: frigate/config.py (hunk 1 above)
Comment (NickM-27) [reply frigate/config.py:820]: > for example, if you add "record" role to input, you doesn't need to additionaly enable recording at camera level

that is absolutely incorrect https://github.com/blakeblackshear/frigate/pull/6444

record must be enabled in the config and added as a role to be turned on

Comment (skrashevich) [reply frigate/config.py:820]: Oh. My apologies

File: frigate/const.py (hunk 2 above)
Comment (NickM-27) [reply frigate/const.py:24]: also in the getting started docs

Screen Shot 2023-06-19 at 12 25 45 PM

Comment (skrashevich) [reply frigate/const.py:24]: > also in the getting started docs

Wrong thread, by you absolutely right. it's my mistake

File: frigate/events/audio.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+        self.detector = AudioTfl()
+        self.shape = (int(round(AUDIO_DURATION * AUDIO_SAMPLE_RATE)),)
+        self.chunk_size = int(round(AUDIO_DURATION * AUDIO_SAMPLE_RATE * 2))
+        self.pipe = f"{CACHE_DIR}/{self.config.name}-audio"
+        self.ffmpeg_cmd = get_ffmpeg_arg_list(
+            FFMPEG_COMMAND.format(
+                " ".join(
+                    self.config.ffmpeg.global_args
+                    + parse_preset_input("preset-rtsp-audio-only", 1)
+                ),
```
Comment (skrashevich) [inline frigate/events/audio.py:139]:
```
            FFMPEG_COMMAND.format(
                " ".join(self.config.ffmpeg.global_args)
                + " ".join(parse_preset_input("preset-rtsp-audio-only", 1)),
```

Comment (NickM-27) [reply frigate/events/audio.py:139]: that won't work, there won't be a space between the end of the first and the beginning of the second groups. Also in general I see nothing wrong with the way it is.

```
>>> print(" ".join(["one", "two"]) + " ".join(["three", "four"]))
one twothree four
```

Comment (skrashevich) [reply frigate/events/audio.py:139]: yeah, it's my copy-paste mistake

```
    def __init__(self, camera: CameraConfig, stop_event: mp.Event) -> None:
        threading.Thread.__init__(self)
        self.name = f"{camera.name}_audio_event_processor"
        self.config = camera
        self.detections: dict[dict[str, any]] = {}
        self.stop_event = stop_event
        self.detector = AudioTfl()
        self.shape = (int(round(AUDIO_DURATION * AUDIO_SAMPLE_RATE)),)
        self.chunk_size = int(round(AUDIO_DURATION * AUDIO_SAMPLE_RATE * 2))
        self.pipe = f"{CACHE_DIR}/{self.config.name}-audio"
        self.ffmpeg_cmd = get_ffmpeg_arg_list(
            self.get_ffmpeg_audio_command("preset-rtsp-audio-only")
        )
        self.pipe_file = None
        self.logpipe = LogPipe(f"ffmpeg.{self.config.name}.audio")
        self.audio_listener = None

    def get_ffmpeg_audio_command(self, preset: str) -> list:
        global_args = " ".join(self.config.ffmpeg.global_args)
        preset_arts = " ".join(parse_preset_input(preset))

        return f"ffmpeg {{}} -i {{}} -f {AUDIO_FORMAT} -ar {AUDIO_SAMPLE_RATE} -ac 1 -y {{}}".format(
            global_args + preset_arts,
            ([i.path for i in self.config.ffmpeg.inputs if "audio" in i.roles][0]),
            self.pipe,
        )
````
(commit https://github.com/skrashevich/frigate/commit/38c1c8fdb4b986ee6a49ec293ff51085c6f8deab)

Comment (skrashevich) [reply frigate/events/audio.py:156]: your version don't work:
```
2023-06-20 10:39:16.347863469  Process audio_capture:
2023-06-20 10:39:16.356396537  Traceback (most recent call last):
2023-06-20 10:39:16.356405545    File "/usr/lib/python3.9/multiprocessing/process.py", line 315, in _bootstrap
2023-06-20 10:39:16.356407064      self.run()
2023-06-20 10:39:16.356408104    File "/usr/lib/python3.9/multiprocessing/process.py", line 108, in run
2023-06-20 10:39:16.356414650      self._target(*self._args, **self._kwargs)
2023-06-20 10:39:16.356415962    File "/opt/frigate/frigate/events/audio.py", line 62, in listen_to_audio
2023-06-20 10:39:16.356418970      audio = AudioEventMaintainer(camera, process_info, stop_event)
2023-06-20 10:39:16.356444799    File "/opt/frigate/frigate/events/audio.py", line 154, in __init__
2023-06-20 10:39:16.356446205      self.config.ffmpeg.global_args
2023-06-20 10:39:16.356447054  TypeError: can only concatenate str (not "list") to str
2023-06-20 10:39:16.401458498  10:39:16.401 DBG [rtsp] new consumer stream=livingroom
2023-06-20 10:39:16.401499738  10:39:16.401 DBG [rtsp] new consumer stream=kitchen
2023-06-20 10:39:16.417735036  10:39:16.417 DBG [streams] start producer url=rtsp://19
```

Comment (NickM-27) [reply frigate/events/audio.py:156]: It does work, I was literally testing audio events with that code for quite a while. Unless there is some platform difference or weird bug with it. I'll take a look.

Comment (skrashevich) [reply frigate/events/audio.py:156]: > It does work, I was literally testing audio events with that code for quite a while. Unless there is some platform difference or weird bug with it. I'll take a look.

Show the minimum example of the working config

Comment (NickM-27) [reply frigate/events/audio.py]: I see, using the default config it works but if global_args are defined then it is a string so it doesn't work.

File: frigate/events/audio.py [hunk 5]
```diff
@@ -166,7 +166,7 @@ def detect_audio(self, audio) -> None:
         if not self.feature_metrics[self.config.name]["audio_enabled"].value:
             return
 
-        logger.error(f"Running audio inference")
+        logger.error("Running audio inference")
```
Comment (skrashevich) [inline frigate/events/audio.py:171]: Error message on any detect_audio call?

Comment (NickM-27) [reply frigate/events/audio.py:171]: Oops thanks, was using it for debugging

File: frigate/events/audio.py [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+    listen()
+
+    for camera in config.cameras.values():
+        if camera.enabled and camera.audio.enabled_in_config:
+            audio = AudioEventMaintainer(camera, process_info, stop_event)
+            audio_threads.append(audio)
+            audio.start()
+
+    while not stop_event.is_set():
+        pass
```
Comment (skrashevich) [inline frigate/events/audio.py:73]: 100% cpu usage in this loop

**Comment by skrashevich** [comment]
Events.jsx:
```
    const filterValues = useMemo(
........
       labels: Object.values(config?.cameras || {})
        .reduce((memo, camera) => {
          memo = memo.concat(camera?.objects?.track || [], camera?.audio?.listen || []); // changed line
          return memo;
        }, config?.objects?.track || [], config?.audio?.listen || []) // changed line
```

**Comment by NickM-27** [comment]
The labels filter is going to be handled in a separate PR since it needs to be reworked for the recent manual_event and attribute label commits

**Comment by skrashevich** [comment]
> The labels filter is going to be handled in a separate PR since it needs to be reworked for the recent manual_event and attribute label commits

But this isn't a reason to violate the integrity of functionality of this PR :)

**Comment by NickM-27** [comment]
> But this isn't a reason to violate the integrity of functionality of this PR :)

The audio events feature works and being able to filter events is not integral to that functionality. 

Blake and I have already discussed the different filters changes that are required after recent PRs, it will be handled separately.

Code reviews are welcome and appreciated, however responses like this are counter productive from my perspective. It is up to Blake what will / won't be accepted.

**Comment by skrashevich** [comment]
I'm extremely surprised by such a sharp reaction and sincerely hope that it is due to a misunderstanding caused by linguistic and cultural differences. 🙄

**Comment by NickM-27** [comment]
I certainly don't mean any disrespect or negativity, tone can easily be misinterpreted especially with linguistic and cultural differences like you said. 

What I was trying to communicate is that from my perspective my original statement 

> The labels filter is going to be handled in a separate PR since it needs to be reworked for the recent manual_event and attribute label commits

explained that that particular functionality would be handled in a separate PR. To then have this statement challenged is tiring.

File: frigate/events/audio.py [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+        now = datetime.datetime.now().timestamp()
+
+        for detection in self.detections.values():
+            if (
+                now - detection.get("last_detection", now)
+                > self.config.audio.max_not_heard
+            ):
+                self.detections[detection["label"]] = None
+                requests.put(
+                    f"http://127.0.0.1/api/events/{detection['event_id']}/end",
```
Comment (skrashevich) [inline frigate/events/audio.py:210]: detection['event_id'] here

File: frigate/events/audio.py [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+        else:
+            resp = requests.post(
+                f"http://127.0.0.1:5000/api/events/{self.config.name}/{label}/create",
+                json={"duration": None},
+            )
+
+            if resp.status_code == 200:
+                event_id = resp.json()[0]["event_id"]
+                self.detections[label] = {
+                    "id": event_id,
```
Comment (skrashevich) [inline frigate/events/audio.py:195]: but detection['id'] here

File: frigate/events/audio.py [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+        for label, score, _ in model_detections:
+            if label not in self.config.audio.listen:
+                continue
+
+            self.handle_detection(label, score)
+
+        self.expire_detections()
+
+    def handle_detection(self, label: str, score: float) -> None:
+        if self.detections.get(label) is not None:
```
Comment (skrashevich) [inline frigate/events/audio.py:182]: `        if self.detections.get(label):`

Comment (NickM-27) [reply frigate/events/audio.py]: Thanks!

File: frigate/events/audio.py (hunk 8 above)
Comment (NickM-27) [reply frigate/events/audio.py:195]: Weird, guess I didn't test long enough to find that, thanks

File: frigate/events/audio.py [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+        now = datetime.datetime.now().timestamp()
+
+        for detection in self.detections.values():
+            if (
+                now - detection.get("last_detection", now)
+                > self.config.audio.max_not_heard
+            ):
+                self.detections[detection["label"]] = None
+                requests.put(
+                    f"http://127.0.0.1/api/events/{detection['id']}/end",
```
Comment (skrashevich) [inline frigate/events/audio.py:210]: `http://127.0.0.1:5000` !

**Comment by blakeblackshear** [comment]
This looks good. We just need the docs updated.

**Comment by NickM-27** [comment]
> This looks good. We just need the docs updated.

Let me know what you think, I figured it didn't make sense to put the object detector and audio detector docs in the same docs page because it could get confusing

**Review by blakeblackshear — APPROVED** [review]

**Comment by RobinBeismann** [comment]
Hey @NickM-27,

thanks for this feature! Any idea when it is going to make it into the stable branch?

**Comment by niemyjski** [comment]
Any reason why there are so many duplicate event names in the list? https://github.com/blakeblackshear/frigate/blob/dev/audio-labelmap.txt

**Comment by NickM-27** [comment]
Because they were a slight variation of that same thing so the name is duplicated so they are called the same thing. 

Makes it easier for users so you don't need kid_yell, adult_yell, etc

**Comment by niemyjski** [comment]
Look at the list, the first 4 items are the same value

**Comment by NickM-27** [comment]
yes, and that is on purpose as I just explained. If you look at the original list they are not the same

**Comment by NickM-27** [comment]
it's the same thing as why `car` is listed twice for the object list https://docs.frigate.video/configuration/objects

**Pull request merged**